### ISSUE TITLE
Options for python

### ISSUE THREAD

**Issue by DavidBruchmann** [webmin/webmin#451]
Currently there are options for perl and PHP but not for python.
Would be nice having overview about modules (beside installation-location) and configuration perhaps.

**Comment by singhdd93** [comment]
+1
Now, its python's time. Webmin/Virtualmin needs configurations options for python now

**Comment by jcameron** [comment]
Interesting idea, but probably isn't going to happen any time soon. We support Perl (because webmin is written in it), and PHP (because it's common).

**Comment by Real-Gecko** [comment]
https://github.com/Real-Gecko/virtualmin-nginx-uwsgi
Try this, not many options available, right now, but it works with Django :D

**Comment by DavidBruchmann** [comment]
Thanks for the link but I think it's too specialized dur to the binding of nginx (even I use it), django and uWSGI.
Due to this combination it's also stated on the page 
"WARNING: This module does not work under Debian 7, making Python 3.2, uWSGI and Django work together under Debian 7 is a real pain."
So it would call many restrictions and probably waste much time for support.

**Comment by Real-Gecko** [comment]
Yeah, Python is not PHP, there're many various ways to make Python app work. I did not try anything other than Django and Python 3, but theoretically it must work with Python 2 and maybe some other framework like Flask.
What are you actual needs?

**Comment by DavidBruchmann** [comment]
I've installed Python 2 and 3 and actually I'm using 2.7 with Sphinx for rendering TYPO3 documents / manuals. TYPO3 itself is programmed in PHP but the extension for rendering is written to use python as sphinx is in Python too. There are still more queries from TYPO3 to other programming languages like perl or java but these are other cases then.
Apart from that surly it's desirable being able to use webmin to configure those things but for a public extension a broader approach should be: reduced functionality (in comparison to your proposition), variability concerning functions and configuration, clear and easy usablity.
These points apply to functionality that might go in a core module, as general "third party modul" naturally any specialism is possible.

**Comment by Real-Gecko** [comment]
So, your python reqeusts are done by calling python directly and not through any kind of gateway?

**Comment by DavidBruchmann** [comment]
Yes and no ;-)
Python is called by php, so network is only indirectly involved.

**Comment by Real-Gecko** [comment]
OK, I see. Does your setup require something like Virtualenv to call some specific python modules?

**Comment by DavidBruchmann** [comment]
Details can be found here:
https://docs.typo3.org/typo3cms/extensions/sphinx/AdministratorManual/InstallingExtension/Index.html
Some dependencies might not be mentioned, so the list may not 100% complete, like sub-modules for Pygments.
But some things depend on the choice in "Choosing the 3rd-party plugins to install".
Furthermore some things are not installed globally but in the TYPO3 directory.

**Comment by DavidBruchmann** [comment]
All right.
I'll have a look at it if i find the time and perhaps a patch will make it
possible.

**Comment by DavidBruchmann** [comment]
right, it's called by php, so network is only indirectly involved.

**Comment by DavidBruchmann** [comment]
Thanks for the link but I think it's too specialized dur to the binding of
nginx (even I use it), django and uWSGI.
Due to this combination it's also stated on the page
"WARNING: This module does not work under Debian 7, making Python 3.2,
uWSGI and Django work together under Debian 7 is a real pain."
So it would call many restrictions and probably waste much time for support.

**Comment by chris001** [comment]
@DavidBruchmann  Here's 3 pages with discussions and practical steps for getting python and apache working well for development on server while not ruining the operating system's dependence on how its own copy of python should remain untouched by other python development on the same system.
https://www.digitalocean.com/community/tutorials/common-python-tools-using-virtualenv-installing-with-pip-and-managing-packages
https://www.digitalocean.com/community/tutorials/how-to-run-django-with-mod_wsgi-and-apache-with-a-virtualenv-python-environment-on-a-debian-vps
https://www.digitalocean.com/community/tutorials/how-to-set-up-python-2-7-6-and-3-3-3-on-centos-6-4

**Comment by DavidBruchmann** [comment]
@chris001 thanks that might be useful as I considered to enable installation in a new python module too!

**Comment by swelljoe** [comment]
Can y'all clarify exactly what "Python support" looks like? The initial post made me think it was about API access to Virtualmin (I guess, since PHP and Perl were mentioned) for Python, but later comments make me think it's about Virtualmin supporting Python web applications in the way it supports things like PHP-FPM, multiple PHP versions, etc. (Which isn't actually true for Perl. Perl support in Virtualmin looks a bit like the Python support...limited, and requires some hand configuration, in many cases, but there is some support.)

In either case, it's a Virtualmin issue, not Webmin (unless you're wanting Webmin API access...but there's no PHP Webmin API, as far as I know, so that doesn't seem to be what folks are wanting here).

My tentative plan looks like:

  - Add support for anyenv (which uses venv behind a unified xenv interface, so we'd only need to support one set of commands across Perl, Ruby, Python, Go, etc.) to Virtualmin, and an awareness of languages and libraries installed in that way. This would allow each virtual server to have its own language environments and versions.
  - Add the ability for users to create a proxy rule (safely and sharing available ports) so they can setup whatever sort of app server they need (every modern web app environment includes their own custom app server...Ruby has Unicorn, Perl has Starman, etc.). This won't be automatic for the near-term, as it would require a huge amount of code for each and every web framework. That's not feasible without a lot more resources than we have (we'd need experts in each of the various languages and environments).

Beyond that, I dunno. Much of the work is pretty specialized, and we already have a lot of code supporting a few dozen PHP apps (and a few in other languages). It's a lot of effort for a small number of users; and those users are probably more knowledgable about how to best deploy their language and framework than we are. So..we'd need a lot of help. Historically, we had someone helping with Python/Django which is why it exists in the install scripts, but we haven't heard from them in a while.

**Comment by chris001** [comment]
It'd be nice to show which python apps/services are running, which are stopped, how much resources are being consumed by them, start/stop button, similar to the services list for apache, mysql, postgresql, proftp, nginx, etc.  On a per  virtual server basis would also be nice.
Just by grepping the process list it ought to be possible to determine which are python apps.
Here are some popular python apps/services to test against.
1. `offlineimap` https://github.com/offlineimap/offlineimap Mail sync app running in python. Runs in one-shot mode or `systemd` daemon mode.
2. `fail2ban` https://github.com/fail2ban/fail2ban this security app runs in daemon mode `systemd` usually.
3. `odoo`  https://github.com/odoo/odoo Business management web apps, python-based, runs on a high port number, usually behind an `nginx` or `apache2` virtual host proxy running for public access on port 80 redirecting to https on port 443.

**Comment by ydaniels** [comment]
any update on this

**Comment by bilekt3** [comment]
Support for python is not good idea. PHP, PERL are different kind of "beasts" than Python. 

Python have more rich ecosystem how to run proper applications (from simple web-server to django framework apps). There are too many options how to start, manage etc. it not as simple as write file index.php and and hit port. I run many Clojure Apps (jars) and some python aps (flask/django) and for me only reverse proxy (nginx) with very specialized configuration is suitable, webmin/virtualmin do great job with many other tasks... And it is not about year 2017 it is about how different technologies works (cgi)...

**Comment by ydaniels** [comment]
Ok understood, I guess using virtualmin + ngnix  (or in addition webmin can support gunicorn) instead of the default apache server would be the best option when considering python but it would be very nice if python apps can enjoy some of the benefit provided by webmin/virtualmin especially when you plan to run php/python project together on the same server. Thanks

**Comment by swelljoe** [comment]
I don't think I entirely agree with @bilekt3 on this one. I think there could be a reasonable implementation that knows how to start a Python app server and proxy to it (either via nginx or Apache). We already do this for node.js and Ruby on Rails apps (though the support is weaker than we'd like...the ecosystems have changed so much so rapidly we just don't have enough people working on it to stay on top of it). In some regards the Python ecosystem is a bit more slow-moving, so that's a benefit for building something that continues to work for more than a couple of months (which is a problem we've had with Ruby and Node as the app servers go in and out of popularity and being maintained so rapidly that it's absurdly difficult, maybe even impossible given that it's mostly just Jamie maintaining it, to target one thing and keep it running on our 9 distro/version combos).

Anyway, the facts are the same as above: We need help. That hasn't changed, and that's the only "update" I can realistically give.

We're not Python experts; I've worked in it professionally in the distant past, but never on web apps. Jamie works with it at his day job, but again not on web apps. If you're using Python as your daily driver for building web apps you already have more expertise in it than we do, and you're better equipped to help design and implement a stack that can be deployed reliably across different distros and both web servers (nginx and Apache is our minimum baseline; we'd rather manage a single app server that can work with either, and using standard OS packages as much as possible), and preferably the same app server for both Flask and Django since supporting both would be great. It may just be updating the existing Django Install Script (I *think* that's in Virtualmin GPL, but it might be Pro only) to provide a bit more flexibility, though there's probably more that people would like.

I mean, even without direct Virtualmin support, someone could create a custom configuration pretty easily using Server Templates to generate the boilerplate web server config, and then use a post-creation script to setup the app server. There's also a [Virtualmin plugin](https://github.com/virtualmin/virtualmin-init) for per-domain startup scripts to handle restarting app servers on reboots. It's not as obvious as deploying a PHP app, but it's not rocket surgery, either.

**Comment by chris001** [comment]
@swelljoe   Agreed,  Probably a good way to start is, get some of the top python apps, and take a look at their setup steps, requirements for running as web (or command line) applications.
To start with, I suggest:
1. https://github.com/odoo/odoo  This is a popular web application running entirely in python.
2. https://github.com/offlineimap/offlineimap  This is a popular command line application running in python, usually either as a daemon process idling and sync'ing every X minutes, or a script that runs it every X minutes.

**Comment by swelljoe** [comment]
I think we're talking about different problems here, @chris001 .

offlineimap is outside of the scope of what I'm talking about. I'm talking exclusively about web apps. You could, I suppose, make a Webmin module to manage offlineimap. There's already fetchmail support in Webmin, which seems like it'd be a similar tool with similar UI and the like if you wanted a model to start from (though the Fetchmail module an old one and nobody uses it, so the code is probably pretty old-fashioned).

Odoo is in scope and looks nice, though. I can't tell if it's a Django app or something else. Do you know what application server it runs under? I'd want to target the sort of obvious execution methods first, and Django and Flask can share the use the same applications types, and they both will run as WSGI apps. If Odoo will, too, then that'd fit into what I have in mind.

**Comment by ydaniels** [comment]
@swelljoe I use odoo regularly, its entirely different from django. Its an ERP system more like SAP, Microsoft Dynamics etc. It can also be used to build web app but its complex than that, i don't think anyone would  use it to build a website unless you are already using it for stuff like HR, Payroll etc. in your company. 
It has a  built-in server which can also be used in production but i think it also support the normal WSGI app mode, infact am not  sure webmin should support something big like this

**Comment by swelljoe** [comment]
It wouldn't be in Webmin (I wish this ticket had been opened on a Virtualmin tracker, as I'm pretty sure everyone here is talking about python app support in Virtualmin...at least, I am). It would be an Install Script in Virtualmin; Webmin doesn't do anything with web apps.

We support several very large web apps in Virtualmin already (Horde, NextCloud, many others, etc.). I'm not saying we *will* support odoo, but if it's a WSGI app, it's at least in the *category* of software we'd consider supporting in Virtualmin.

**Comment by chris001** [comment]
The production deployment for odoo is here:
https://www.odoo.com/documentation/10.0/setup/deploy.html
It looks like the most secure higgest performing option for an odoo web app facing the internet is behind an nginx HTTPS reverse proxy.  Full configuration is given.
They explain odoo WSGI deployment, compared with nginx reverse HTTPS proxy, WSGI is more awkward in its support of the long polling live chat server and the cron workers.

Note: odoo 11.0 is going to be released at the Odoo Experience Days conference on Thursday Oct 5 - five days from now.
https://www.odoo.com/groups/community-59/community-27322288
https://www.odoo.com/event/odoo-experience-2017-2017-10-04-2017-10-06-692/page/oxp17-homepage

**Comment by swelljoe** [comment]
OK, that sounds like it's too complicated for an Install Script (and probably too complicated and requiring too much privilege, for any shared hosting environment). I did a little googling, and it sounds like someone would need root to be able to manage an Odoo installation. So, we probably have to rule that one out as well. I mean, there's nothing stopping someone with root privileges from installing odoo in a Virtualmin environment, but I don't see any easy way to make it safe in a shared hosting context.

My short-term goal is to make it easier for app servers (of any type) to be configured behind a proxy (whether apache or nginx; whatever is the web server in use on the system) by Virtualmin domain owners, which would allow any application that needs an app server (which includes Django/Flask apps, Node.js apps, Ruby on Rails apps, etc.) to be configured without root privileges and without accidentally (or intentionally) stepping on other domains on the system. Jamie added protection for port-stealing in 6.0, and we already setup proxy rules for Django, RoR and Node apps in install scripts (but not in a general way), so we're about halfway there, I think.

**Comment by chris001** [comment]
@swelljoe 
Here's the odoo 10 community (free) + enterprise (paid) install script.
https://www.odoo.yenthevg.com/installing-odoo-10-ubuntu-16-04/
The script can install multiple odoo instances in one server, because it can set each instance to use a different port.
As far as root privileges, the script might need to install missing packages, which requires root, and create a postgresql user and postgresql database, which might require root also.  Also, the script sets permissions for security reasons, which also might require root.  All these reasons for requiring root are legit and standard practice, so it should be OK for virtualmin to elevate to root to install odoo with this script.

**Comment by bilekt3** [comment]
@chris001 could you please elaborate on 

> All these reasons for requiring root are legit and standard practice

I cannot find legit reason why web server need this type of privileges. I am not Odoo expert, but for me is still running web server after nginx. (This is not offensive, I am trying to understand situation here)

@swelljoe Thanks for feedback, I haven't luck with server templates for init reverse proxy, could you please five me direction for some docs? It might be valuable to create support document for Python (or generally all reverse proxy deploy scenarios - Clojure, Node.js etc.) for all Virtualmin users. Which I would like to draft.

**Comment by chris001** [comment]
@bilekt3 
The Odoo install script must install missing system packages, which requires root.  And configure postgresql which might require root also if it is system-wide configuration of postgresql. Also, the Odoo installer script creates the `odoo` user and home folder, and sets permissions on the`/odoo/odoo-server` folder for security reasons, which also might require root. to do that

**Comment by lisandi** [comment]
@ydaniels
we use Odoo 8 9 10 and now 11 and in some days also Flectra on Ubuntu 17.10 with Virtualmin/Webmin and it works just perfect. Our customers run mostly websites with it ;-) That is the main purpose since Odoo 7 as they like to have their ERP stuff in the same app like their website as a one stop solution. The odoo website builder is a wysiwyg solution whch until now isn't reached even by those wordpress website builders. In short it would be great to have Python support available.

NGINX is not really needed as most other apps run with apache which gets used for the Odoo installations as proxy. All Odoo instances run in a virtualmin virtualserver under /home/user/odoo-server (which holds the odoo code)
and odoo-custom which holds all custom modules and themes 
the configuration files are located in etc/odoo and etc/init.d as well as in lib/systemd/system/odoo-server.service
each server has its own user which we create with virtualmin

steps:
we create a usual virtual server user
we activate the apache for a certain domains or when more than one site runs from the same instance we create more domains under the same user
Odoo can manage many installations with one core by using a database filter which we use. 
The filter is the domain name.
The source gets installed under the specific user and we create manually the custom folder 
it is fast installed and simply works.

We start stop the python servers by using the bootup and shutdown section in virtualmin.

Those missing packages also need to be installed with any other ruby or whatever application which is not using the standard php set. Even TYPO3 needs additional server parts ;-) and that is a PHP application - also requiring root. for customers of virtualmin this means nothing else as that they need to contact the support and ask them to install the missing stuff. or simply change the hoster to one who has already everything installed like us who run TYPO3, Drupal, WordPress, Odoo, Flectra and Ruby stuff on the same server

Odoo does a great job actually by itself as it holds a requirements.txt and installs all missing parts itself!

**Comment by chris001** [comment]
IMHO this python app management feature would be best implemented as 
1. Have virtualmin be able to create a separate python `virtualenv` (virtual environment) for any python web app or console app. The purpose of which is to let the app run, and to run, it often needs to get around the problem of each app's required python package versions, often conflict and are incompatible with the "system python" package versions.
2, start/stop/restart/reload control panel buttons, and python web service process health status icons.
3. One-time proxy setup and teardown. In case the python web app expects a person to setup a proxy manually. In fact, virtualmin can be made smart enough to setup `nginx` and `apache2` proxies in front of a python web service, per domain or subdomain name. Plus a free TLS cert from Let's Encrypt.

**Comment by lisandi** [comment]
Very good described, virtualenv is also needed as most python apps run in 2.7 others already in 3.6 running both would be possible using vitualenv

**Comment by chris001** [comment]
@lisandi 
Exactly.  Ubuntu 16.04 system distro packaged python version is 3.5.  Debian Jessie 8 system distro packaged python version is 2.7.9.  Many mature python apps (e.g. offlineimap console app) run "best" on python 2.7,x.  Many newer python apps (odoo 11.0 web service app) need python 3.5 minimum. `virtualenv` is necessary for both apps running on the same server without causing malfunctions due to overwriting the global python libraries with versions that are either too old or too new for the other python apps running on the system.

**Comment by chris001** [comment]
Working a lot in python the last few months, and it's become more clear, to support managing installing python web apps for "production" and "development" is something webmin/virtualmin should probably do.  It's very doable. You only need to decide a few parameters:
1. version of python to run it under
2. which web server to use. production must use a real web server: nginx or apache2, with LE TLS certs ideally, and a real domain.  development can get away with a smaller dev-only web server.
3. create a `virtualenv`, enter the `virtualenv`. either `git` download the app and `pip install -r requirements.txt`, or `pip install` the app directly if it's an established app on `PyPI`. 
4. settings for WSGI connection between the web server and the python app, thru the config for the hostname in site config file, 
5. add `systemd` service file for the python app, so it starts as daemon on system start.
6. setting for path to log file.
7. display start/stop/restart/status like all other web services.
8. to remove/uninstall: stop the python service, remove the dir containing the `virtualenv`, remove `systemd` service, remove web server site config, reload daemon configs for web server and `systemd`.

**Comment by chris001** [comment]
Working a lot in python the last few months, and it's become more clear, to support managing installing python web apps for "production" and "development" is something webmin/virtualmin should probably do.  It's very doable. You only need to decide a few parameters:
1. version of python to run it under
2. which web server to use. production must use a real web server: nginx or apache2, with LE TLS certs ideally, and a real domain.  development can get away with a smaller dev-only web server.
3. create a `virtualenv`, enter the `virtualenv`. either `git` download the app and `pip install -r requirements.txt`, or `pip install` the app directly if it's an established app on `PyPI`. 
4. settings for WSGI connection between the web server and the python app, thru the config for the hostname in site config file, 
5. add `systemd` service file for the python app, so it starts as daemon on system start.
6. setting for path to log file.
7. display start/stop/restart/status like all other web services.
8. to remove/uninstall: stop the python service, remove the dir containing the `virtualenv`, remove `systemd` service, remove web server site config, reload daemon configs for web server and `systemd`.

**Comment by stefanlk** [comment]
I think, in debian bullseye, there is no standard command "/usr/bin/python" anymore.
I tried to use the tool text login and get the error "command python not found".

**Comment by swelljoe** [comment]
@stefanlk that's a wholly independent issue. I know this ticket is stupidly vague throughout, and it seems to be a dumping ground for anybody with a Python-related wish/question/issue that might be have some sort of tenuous connection to Webmin or Virtualmin. But, that kind of vaguery is completely useless for actually getting anything done, it's just a bunch of talking.

So, can you open a new issue with your specific problem? Your problem can be solved easily, I think, unlike most of the big conceptual stuff being talked about in this issue.

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
